src/app/qgsmaptoolfeatureaction.cpp

```diff
@@ -121,36 +121,67 @@ bool QgsMapToolFeatureAction::doAction( QgsVectorLayer *layer, int x, int y )
     QgsDebugMsg( QStringLiteral( "Caught CRS exception %1" ).arg( cse.what() ) );
   }
 
-  QgsAction defaultAction = layer->actions()->defaultAction( QStringLiteral( "Canvas" ) );
-
+  QgsFeature f;
+  QgsFeatureList features;
   QgsFeatureIterator fit = layer->getFeatures( QgsFeatureRequest().setFilterRect( r ).setFlags( QgsFeatureRequest::ExactIntersect ) );
-  QgsFeature feat;
-  while ( fit.nextFeature( feat ) )
+  while ( fit.nextFeature( f ) )
+  {
+    features.append( f );
+  }
+
+  if ( !features.isEmpty() )
   {
-    if ( defaultAction.isValid() )
+    if ( features.count() == 1 )
     {
-      // define custom substitutions: layer id and clicked coords
-      QgsExpressionContext context;
-      context << QgsExpressionContextUtils::globalScope()
-              << QgsExpressionContextUtils::projectScope( QgsProject::instance() )
-              << QgsExpressionContextUtils::mapSettingsScope( mCanvas->mapSettings() );
-      QgsExpressionContextScope *actionScope = new QgsExpressionContextScope();
-      actionScope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "click_x" ), point.x(), true ) );
-      actionScope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "click_y" ), point.y(), true ) );
-      actionScope->addVariable( QgsExpressionContextScope::StaticVariable( QStringLiteral( "action_scope" ), QStringLiteral( "Canvas" ), true ) );
-      context << actionScope;
-
-      defaultAction.run( layer, feat, context );
+      doActionForFeature( layer, features.first(), point );
     }
     else
     {
-      QgsMapLayerAction *mapLayerAction = QgsGui::mapLayerActionRegistry()->defaultActionForLayer( layer );
-      if ( mapLayerAction )
+      QMenu *featureMenu = new QMenu();
+      for ( int idx = 0; idx < features.count(); idx++ )
```

Comment:
Could that be done with an iterator based approach ( `for ( const QgsFeature &feature : features )`)?